java/arcs/core/host/ParticleStateMachine.kt

```diff
@@ -0,0 +1,172 @@
+                        particle.handles.reset()
+                    }
+                }
+
+                onError(::markParticleAsFailed)
+
+                on(StartEvent, Created)
+                on(FailedEvent, Failed)
+            }
+
+            state(Created) {
+                onEntry {
+                    val waitingSet = mutableSetOf<Handle>()
+                    waitingSet.addAll(handles.values)
+
+                    handles.values.forEach { handle ->
+                        handle.onReady {
+                            waitingSet.remove(handle)
+                            if (waitingSet.isEmpty()) {
+                                GlobalScope.launch {
+                                    triggerImmediate(ReadyEvent)
+                                }
+                            }
+                        }
+                    }
+                }
+
+                onError(::markParticleAsFailed)
+
+                on(ReadyEvent, Ready)
+                on(FailedEvent, Failed)
+            }
+
+            state(Ready) {
+                onEntry {
+                    particle.onReady()
+                    trigger(StartEvent)
+                }
+
+                onError(::markParticleAsFailed)
+
+                on(FailedEvent, Failed)
+                on(StartEvent, Started)
+            }
+
+            state(Failed) {
+                onEntry(checkForMaxFailures)
+                onError {
+                    trigger(MaxFailedEvent)
+                }
+                on(MaxFailedEvent, MaxFailed)
+                on(StartEvent, Created)
+                on(StopEvent, Failed)
+            }
+
+            state(Failed_NeverStarted) {
+                onEntry(checkForMaxFailures)
+                onError {
+                    trigger(MaxFailedEvent)
+                }
+                on(StartEvent, Instantiated)
+                on(MaxFailedEvent, MaxFailed)
+                on(StopEvent, Failed_NeverStarted)
+            }
+
+            state(MaxFailed) {
+                // There is no escape
+            }
+        }
+    }
+
+    private val checkForMaxFailures: suspend (ParticleContext).() -> Unit = {
+        consecutiveFailureCount++
+        if (consecutiveFailureCount >= MAX_CONSECUTIVE_FAILURES) {
+            // Throwing clears the pending event queue
+            throw MaxFailuresException()
+        }
+    }
+
+    class MaxFailuresException : IllegalStateException()
+
+    companion object {
+        /**
+         * Move to [ParticleState.Failed] if this particle had previously successfully invoked
+         * [Particle.onCreate()], else move to [ParticleState.Failed_NeverStarted]. Increments
+         * consecutive failure count, and if it reaches maximum, transitions to
+         * [ParticleState.MaxFailed].
+         */
+        private suspend fun markParticleAsFailed(
+            particleContext: ParticleContext,
+            exception: Exception
+        ) = particleContext.run {
```

Comment:
why is `.run` needed?  

If this method needs to return Unit, it's cleaner/more readable to just do:

```kotlin
private suspend fun markParticleAsFailed(
    particleContext: ParticleContext,
    exception: Exception
) {
    particleContext.trigger(FailedEvent)
}
```